core/include/scipp/core/element/to_unit.h

```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#pragma once
+
+#include "scipp/common/overloaded.h"
+#include "scipp/core/element/arg_list.h"
+#include "scipp/core/time_point.h"
+#include "scipp/core/transform_common.h"
+
+namespace scipp::core::element {
+
+namespace {
+template <class T>
+constexpr auto round =
+    [](const auto x) { return static_cast<T>(x < 0 ? x - 0.5 : x + 0.5); };
+}
+
+constexpr auto to_unit = overloaded{
+    arg_list<double, std::tuple<float, double>, std::tuple<int64_t, double>,
+             std::tuple<int32_t, double>, std::tuple<time_point, double>>,
+    transform_flags::expect_no_variance_arg<1>,
+    [](const units::Unit &, const units::Unit &target) { return target; },
+    [](const auto &x, const auto &scale) {
+      using T = std::decay_t<decltype(x)>;
+      if constexpr (std::is_same_v<T, time_point>)
+        return T{round<int64_t>(x.time_since_epoch() * scale)};
+      else if constexpr (std::is_integral_v<T>)
+        return round<T>(x * scale);
+      else
+        return static_cast<T>(x * scale);
+    }};
```

Comment:
Could be implemented with `underlying` similar to https://github.com/scipp/scipp/blob/50a0d4b9a6580a1cdcd2860d81b540ba04232421/core/include/scipp/core/element/creation.h#L43

Then, this code would only have to decide whether to round or not.

Reply:
Well, not quite, since we don't just have to extract the int from the time_point, but also put it back. I had started using `underyling` (moved to a new file), but it ended up more messy, since what is being rounded to is not what we want to return.